File: samples/java/src/main/java/com/example/event/BankAccountEvent.java
// Copyright © 2017-2021 UKG Inc. <https://www.ukg.com>

package com.example.event;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

import java.util.UUID;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME,
        property = "name")
@JsonSubTypes({
        @JsonSubTypes.Type(value = BankAccountCreated.class, name = "BankAccountCreated"),
        @JsonSubTypes.Type(value = BankAccountUpdated.class, name = "BankAccountUpdated"),
})
public interface BankAccountEvent {

    UUID getAccountNumber();
}

File: samples/scala/src/main/scala/com/example/event/BankAccountEvent.scala
package com.example.event

import play.api.libs.json.{Format, JsValue, Json}

import java.util.UUID

sealed trait BankAccountEvent {
  def accountNumber: UUID
  def toJson: JsValue
}

object BankAccountCreated {
  implicit val format: Format[BankAccountCreated] = Json.format
}
case class BankAccountCreated(accountNumber: UUID, accountOwner: String, securityCode: String, balance: Double) extends BankAccountEvent {
  override def toJson: JsValue = Json.toJson(this)
}

object BankAccountUpdated {
  implicit val format: Format[BankAccountUpdated] = Json.format
}
case class BankAccountUpdated(accountNumber: UUID, newBalance: Double) extends BankAccountEvent {
  override def toJson: JsValue = Json.toJson(this)
}